### DISCUSSION TITLE
Capture navigation events

### DISCUSSION THREAD

**Discussion by HakanL** [unoplatform/uno.extensions#1588] [Q&A] [answered]
I have a project where I'm using external navigation controls (basically a rotary encoder that you twist and click to navigate the controls on the screen). With that I need to interact on a global app-level with the current view model (the list of controls that I can navigate between is controlled, including the order of which it should be navigated in). For it to be efficient it would be nice to be able to hook into the navigation requests so I can tell when the user navigates between pages. The `IRouteUpdater` seems like a perfect candidate, but it's an internal interface so I can't replace it in my service collection. Any suggestion on how to do this? I may have an incorrect approach, but my question may still be valid, how I can determine on an app-level what navigations happen.

**Comment by nickrandolph** [comment, marked as answer]
You're right, the IRouteUpdater is internal but the [IRouteNotifier](https://github.com/unoplatform/uno.extensions/blob/main/src/Uno.Extensions.Navigation.UI/IRouteNotifier.cs) is public. You should be able to request the implementation of this interface and attach an event handler. 

I'm not sure I quite follow what you want to be able to do when you intercept the navigation event, so if you can provide more context we can either provide guidance or expose different aspects of navigation if needed

**Reply by HakanL** [reply]
Ah I see, let me look at IRouteNotifier if that may work. Yeah, it's a little tricky to explain, I'm building a menu system with multiple levels (different data in the same VM) and I have a set list of controls in the VM that the focus (rotary encoder) will cycle through. So I need to know what my current VM is so I know what elements are available to the focus on. I know there's the FocusManager, but the "NextControl" method is not consistent, so I need to manually tell it what control to focus on when you twist the knob left or right. The inconsistency with the FocusManager is that the main Tab/Shift-Tab uses Next/Previous, but when you're inside a list view then it's Up/Down/Left/Right. I only have "twist left" and "twist right", so I need to move the focus through all available controls using only that.

**Reply by HakanL** [reply]
@nickrandolph `IRouteNotifier` worked great, only weird thing is when I display a dialog then the notify is of the parent page/view, and when the dialog is closed then basically null is in the event args (Region is null, etc).

**Reply by nickrandolph** [reply]
Oh interesting I might take a look and see if we can adjust that

**Reply by HakanL** [reply]
@nickrandolph It would be nice if the IRouteNotifier event could get the details of the dialog.

**Reply by nickrandolph** [reply]
@HakanL do you need a reference to the dialog in the IRouteNotifier when the dialog closes?
Currently the IRouteNotifier returns the root region for the hierarchy where navigation commences. As dialogs are disconnected (since modal) from the main navigation hierarchy, this is why you get a null View when dialog is closed (the view doesn't exist any more)

**Reply by HakanL** [reply]
When the dialog closes I care less about the dialog that is closed, and more about what is now displayed. Of course more is always better, but I assume it will be less changes if I can just get the "previous page". I was able to hack my way through it by implementing my own INavigator (which I also use to dispose VMs when they are no longer in the current stack, I subscribe to external events so I need to know when a VM/page will no longer be resurrected so I can unsubscribe).

**Reply by nickrandolph** [reply]
Sounds like you've dug deep into navigators - can I make a request that you raise an issue with all the extra things you're doing in your INavigator. Where possible we'd like to review and incorporate those that make sense (ie are generally applicable to other apps).

With regards to IRouteNotifier, the difficulty is that the RouteChanged event gets raised for the hierarchy where the navigation is triggered. When opening a dialog this is the main hierarchy of the app; When closing the dialog this is the separate hierarchy of the dialog. One option would be to raise two RouteChanged events, one for each hierarchy - does this sound reasonable?

The other option would be to extend the NavigationResponse to give you access to the Region and/or Navigator where the navigation was executed. This sounds easy but there are some scenarios where this may be multiple regions (for example tabs where there is a navigator for the tabbar and a navigator for the grid where the content is displayed).

**Reply by HakanL** [reply]
I can certainly try to write up with I need and raise as an issue. On a high level I'm trying to accomplish two things:

**Lifetime control**
Since some of my viewmodels subscribe (RX) to external events (background player updates, etc) I need to unsubscribe (IDisposable) to not have 100s of unused subscriptions triggering every time there's an update for view models that aren't in the navigation stack any more. For example, I have a timer in my base view model that updates a string with the current time once per second. Once I've navigated around my application I may have 20 of these timers running (since there's no built-in "event" to tell the VM that it's no longer in-use. Using the IRouteNotifier I can certainly know when the VM isn't "current" (except for when a dialog is shown), but the issue is that the user may navigate back in the stack, and then the VM is just resurrected (it's not re-created), at which point I need the timer to run. What I ended up doing in my INavigator was to just keep track of the VMs (and pages) that are in the stack and dispose of them when they can't be navigated to (without being re-created).
What would be helpful would be if the built-in Uno navigator would dispose VMs/pages if they implement IDisposable once they are out of navigation-stack.
Another related feature that would help the issue of event subscriptions in VMs would be if there was an interface for let's call is hydration/dehydration. So any time a VM/page is out of view (but still in the nav stack) then it could call a Dehydrate method on the class. It could then turn off subscriptions/timers. Then if the user navigates back to the same view then the hydration method is called and the timers could be restarted.

**External navigation**
As I've described above I have a rotary encoder and push button on my RPi that the user can use to navigate around the screen (if they don't feel like using the touch screen). So any time there's a new view on the screen the "external focus manager" has to know what's on the screen so it can keep a list of what elements to focus on (it's similar to keyboard focus, but each view can dictate which elements can be focused and in which order). So this is where the IRouteNotifier has been very helpful, except for when dialog boxes are shown, since they may have focus elements, but my focus manager isn't aware of them (IRouteNotifier is triggered with the calling view, not the dialog). I had to walk the visual tree to find the dialog, which works, but it would be much cleaner to be properly notified. Could I have solved it my just using the FocusManager and emulated keyboard? Perhaps, but I ran into problems where some elements used up/down/left/right on the keyboard once it has focus, and others are standard Tab/Shift-tab. Having just a single knob that can be rotated left or right made that hard to map in a generic way.

I hope that helps, I don't necessarily have a complete picture how this could be implemented in Uno itself, but those are issues I've tried to solve (and I'll be the first one to acknowledge that my MVVM skills are not my strongest).

**Reply by nickrandolph** [reply]
Here's the update that was made - https://github.com/unoplatform/uno.extensions/pull/1678
Let me know whether this solves your scenarios - currently this is only in the 3.0 release

**Reply by HakanL** [reply]
Looks good, I'll see if I can upgrade to 3.0 later and see if this makes it simpler for what I'm attempting to do. Thanks!
